In [1]:
# Running the Colab Drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


#Importing of the necessary libraries and the modules

In [2]:
import numpy as np
import pandas as pd
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score
from sklearn import preprocessing
from sklearn.model_selection import GridSearchCV
!pip install xgboost
import xgboost as xgb
from xgboost import XGBClassifier
from sklearn.model_selection import cross_val_score, StratifiedKFold
from sklearn.metrics import roc_auc_score

from xgboost.sklearn import XGBClassifier


#Loading data

In [3]:
data = pd.read_csv('/content/drive/MyDrive/Home Credit/home-credit-default-risk/dropped_train.csv').sample(frac =.50)
data

TARGET  CNT_CHILDREN  ...  income_regularity  balance_regularity
202498     0.0           0.0  ...                NaN                 NaN
130684     0.0           0.0  ...                NaN                 NaN
47770      0.0           0.0  ...                NaN                 NaN
154656     0.0           1.0  ...                NaN                 NaN
266524     0.0           1.0  ...                NaN                 NaN
...        ...           ...  ...                ...                 ...
140506     0.0           1.0  ...                NaN                 NaN
33340      0.0           1.0  ...                NaN                 NaN
135641     0.0           2.0  ...                NaN                 NaN
259209     0.0           0.0  ...                NaN                 NaN
14389      0.0           1.0  ...                inf                 NaN

[153756 rows x 856 columns]

#Train Test Split

In [ ]:
X = data.drop(['TARGET'], axis=1)
y = data['TARGET']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.30, random_state =42)


In this problem, the data is imbalanced. So we can’t use accuracy as a error metric. When data is imbalanced we can use Log loss, F1-score and AUC. Here we are sticking to AUC which can handle imbalanced datasets.

Due to the high unbalance between 0 and 1 target flags, we set  the scale pos weight parameter  to the ratio between the number of 0s and the number of 1s. 

Area Under Curve (AUC)

An ROC curve is the most commonly used way to visualize the performance of a binary classifier, and AUC is (arguably) the best way to summarize its performance in a single number.

In [ ]:
ratio = (y == 0).sum()/ (y == 1).sum()
ratio

11.447862694300518

#Model XGBoost

In [ ]:
# fit xgboost on an imbalanced classification dataset
from numpy import mean
from sklearn.datasets import make_classification
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
clf = XGBClassifier(n_estimators=1000, objective='binary:logistic', gamma=0.1, subsample=0.5, scale_pos_weight=ratio )
clf.fit(X_train, y_train, eval_set=[(X_test, y_test)], eval_metric='auc', early_stopping_rounds=10)

[0]	validation_0-auc:0.681304
Will train until validation_0-auc hasn't improved in 10 rounds.
[1]	validation_0-auc:0.692027
[2]	validation_0-auc:0.695133
[3]	validation_0-auc:0.70142
[4]	validation_0-auc:0.704452
[5]	validation_0-auc:0.70766
[6]	validation_0-auc:0.70839
[7]	validation_0-auc:0.710084
[8]	validation_0-auc:0.712798
[9]	validation_0-auc:0.713794
[10]	validation_0-auc:0.715551
[11]	validation_0-auc:0.718352
[12]	validation_0-auc:0.721639
[13]	validation_0-auc:0.724157
[14]	validation_0-auc:0.726897
[15]	validation_0-auc:0.72826
[16]	validation_0-auc:0.729489
[17]	validation_0-auc:0.731814
[18]	validation_0-auc:0.734178
[19]	validation_0-auc:0.736313
[20]	validation_0-auc:0.737915
[21]	validation_0-auc:0.738592
[22]	validation_0-auc:0.739411
[23]	validation_0-auc:0.740481
[24]	validation_0-auc:0.741897
[25]	validation_0-auc:0.742635
[26]	validation_0-auc:0.74316
[27]	validation_0-auc:0.744308
[28]	validation_0-auc:0.744449
[29]	validation_0-auc:0.745409
[30]	validation_0-auc

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0.1,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=1000, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=11.447862694300518,
              seed=None, silent=None, subsample=0.5, verbosity=1)

In [ ]:
print("best score: {0}, best iteration: {1}, best ntree limit {2}".format(clf.best_score, clf.best_iteration, clf.best_ntree_limit))

best score: 0.777008, best iteration: 204, best ntree limit 205


#Best Metrics

In [ ]:
n_estimators = clf.best_ntree_limit
clf = XGBClassifier(n_estimators=n_estimators, objective='binary:logistic', seed=27, gamma=0.1, max_depth=4, subsample=0.8, nthread=4, colsample_bytree=0.8, min_child_weight=4,  scale_pos_weight=2)
clf.fit(X_train, y_train, eval_set=[(X_test, y_test)], eval_metric='auc', early_stopping_rounds=10)

[0]	validation_0-auc:0.693295
Will train until validation_0-auc hasn't improved in 10 rounds.
[1]	validation_0-auc:0.700719
[2]	validation_0-auc:0.70923
[3]	validation_0-auc:0.71326
[4]	validation_0-auc:0.719422
[5]	validation_0-auc:0.723466
[6]	validation_0-auc:0.724168
[7]	validation_0-auc:0.72482
[8]	validation_0-auc:0.725757
[9]	validation_0-auc:0.726074
[10]	validation_0-auc:0.726654
[11]	validation_0-auc:0.728441
[12]	validation_0-auc:0.728905
[13]	validation_0-auc:0.730304
[14]	validation_0-auc:0.730067
[15]	validation_0-auc:0.731447
[16]	validation_0-auc:0.733094
[17]	validation_0-auc:0.733734
[18]	validation_0-auc:0.734048
[19]	validation_0-auc:0.735963
[20]	validation_0-auc:0.737659
[21]	validation_0-auc:0.738807
[22]	validation_0-auc:0.739625
[23]	validation_0-auc:0.74003
[24]	validation_0-auc:0.742082
[25]	validation_0-auc:0.743108
[26]	validation_0-auc:0.744706
[27]	validation_0-auc:0.745644
[28]	validation_0-auc:0.747046
[29]	validation_0-auc:0.747254
[30]	validation_0-au

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=0.8, gamma=0.1,
              learning_rate=0.1, max_delta_step=0, max_depth=4,
              min_child_weight=4, missing=None, n_estimators=205, n_jobs=1,
              nthread=4, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=2, seed=27,
              silent=None, subsample=0.8, verbosity=1)

In [ ]:
# y_pred = xgb_model.predict(X_test)
y_pred = clf.predict(X_test)

#Results

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

         0.0       0.93      0.99      0.96     42462
         1.0       0.43      0.13      0.19      3665

    accuracy                           0.92     46127
   macro avg       0.68      0.56      0.58     46127
weighted avg       0.89      0.92      0.90     46127



After using best_ntree_limit, we increase our ROC-AUC value from 0.777 to 0.7798. However, this result is remarkably lower than LightGBM. Lets continue to LightGBM.